python/dgl/function/message.py

```diff
@@ -25,233 +27,193 @@ def name(self):
-
-    Only scalar feature is supported currently.
-    """
-    feat = g.get_e_repr()[field]
-    shape = F.shape(feat)
-    return len(shape) == 1 or (len(shape) == 2 and shape[1] == 1)
 
-
-class SrcMulEdgeMessageFunction(MessageFunction):
-    """Class for the src_mul_edge builtin message function.
+class BinaryMessageFunction(MessageFunction):
+    """Class for the lhs_op_rhs builtin message function.
 
     See Also
     --------
     src_mul_edge
     """
-    def __init__(self, mul_op, src_field, edge_field, out_field):
-        self.mul_op = mul_op
-        self.src_field = src_field
-        self.edge_field = edge_field
+    def __init__(self, binary_op, lhs, rhs, lhs_field, rhs_field, out_field):
+        self.binary_op = binary_op
+        self.lhs = lhs
+        self.rhs = rhs
+        self.lhs_field = lhs_field
+        self.rhs_field = rhs_field
         self.out_field = out_field
 
-    def is_spmv_supported(self, g):
-        """Return true if this supports SPMV optimization.
-
-        Parameters
-        ----------
-        g : DGLGraph
-            The graph.
-
-        Returns
-        -------
-        bool
-            True if this supports SPMV optimization.
-        """
-        return _is_spmv_supported_edge_feat(g, self.edge_field)
-
-    def __call__(self, edges):
-        """Regular computation of this builtin function
-
-        This will be used when optimization is not available and should
-        ONLY be called by DGL framework.
+    def __call__(self, graph, src_frame, dst_frame, edge_frame, out_size,
+                 src_map, dst_map, edge_map, out_map, reducer="none"):
+        """Symbolic computation of builtin binary message function to create
+        runtime.executor
         """
-        sdata = edges.src[self.src_field]
-        edata = edges.data[self.edge_field]
-        # Due to the different broadcasting semantics of different backends,
-        #   we need to broadcast the sdata and edata to be of the same rank.
-        rank = max(F.ndim(sdata), F.ndim(edata))
-        sshape = F.shape(sdata)
-        eshape = F.shape(edata)
-        sdata = F.reshape(sdata, sshape + (1,) * (rank - F.ndim(sdata)))
-        edata = F.reshape(edata, eshape + (1,) * (rank - F.ndim(edata)))
-        ret = self.mul_op(sdata, edata)
-        return {self.out_field : ret}
+        graph = var.GRAPH(graph)
+        in_frames = (src_frame, dst_frame, edge_frame)
+        in_maps = (src_map, dst_map, edge_map)
+        lhs_data = ir.READ_COL(in_frames[self.lhs], var.STR(self.lhs_field))
+        rhs_data = ir.READ_COL(in_frames[self.rhs], var.STR(self.rhs_field))
+        lhs_map = var.MAP(in_maps[self.lhs])
+        rhs_map = var.MAP(in_maps[self.rhs])
+        out_map = var.MAP(out_map)
+        return ir.BINARY_REDUCE(reducer, self.binary_op, graph, self.lhs,
+                                self.rhs, lhs_data, rhs_data, out_size,
+                                lhs_map, rhs_map, out_map)
```

Comment:
Add a TODO here: ideally `dgl.function` shouldn't depend on IR.

Reply:
Hmm... Unfortunately, this is a "feature"... Because we want to unify all DGL kernel interfaces to support more kernels, the message function interface has to be unified. I designed this so that scheduler does not need knowledge about the what builtin message and reduce function are. If you think this is problematic, we probably need to discuss if other choices are available.

If you just don't like my using __call__ because user might use it as well, I can instead use some other private function name for the symbolic execution

Reply:
> If you just don't like my using **call** because user might use it as well, I can instead use some other private function name for the symbolic execution

Yeah let's do that.

Reply:
Done